crates/red_knot_python_semantic/src/types.rs

```diff
@@ -55,37 +60,40 @@ fn symbol_ty_by_id<'db>(db: &'db dyn Db, scope: ScopeId<'db>, symbol: ScopedSymb
         let declarations = use_def.public_declarations(symbol);
         // If the symbol is undeclared in some paths, include the inferred type in the public type.
         let undeclared_ty = if declarations.may_be_undeclared() {
-            Some(bindings_ty(
-                db,
-                use_def.public_bindings(symbol),
-                use_def
-                    .public_may_be_unbound(symbol)
-                    .then_some(Type::Unbound),
-            ))
+            Some(
+                bindings_ty(db, use_def.public_bindings(symbol))
+                    .map(|bindings_ty| Symbol::Type(bindings_ty, use_def.public_boundness(symbol)))
+                    .unwrap_or(Symbol::Unbound),
+            )
         } else {
             None
         };
         // Intentionally ignore conflicting declared types; that's not our problem, it's the
         // problem of the module we are importing from.
-        declarations_ty(db, declarations, undeclared_ty).unwrap_or_else(|(ty, _)| ty)
+        match undeclared_ty {
+            Some(Symbol::Type(ty, boundedness)) => Symbol::Type(
+                declarations_ty(db, declarations, Some(ty)).unwrap_or_else(|(ty, _)| ty),
+                boundedness,
```

Comment:
I think what boundness value we should apply to a symbol that is declared in some paths (but not all) and bound in some paths (but not all) is pretty tricky and requires further attention in future. Right now here we are always just using the boundness from bindings and ignoring declarations for purposes of boundness, but that's weird, because we only even look at bindings if the symbol may be undeclared, which means in this example:

```py
x: int

if flag:
    y: int
else
    y = 3
```

If we import from this module, we will currently report `x` as a definitely-bound symbol (even though it has no bindings at all!) but report `y` as possibly-unbound (even though every path has either a binding or a declaration for it.)

Possibly the behavior here should differ depending on whether we are importing from a stub or a regular Python module? But an argument could be made that even in a regular Python module, if it declares `x: int` at module level, we should just trust that and not require that we can see the binding of it.

This is not for this PR, but it may be worth a TODO comment here.